File: spring-meilisearch/src/main/java/com/github/asm0dey/opdsko_spring/meilisearch/MeilisearchDockerComposeConnectionDetailsFactory.java
package com.github.asm0dey.opdsko_spring.meilisearch;

import org.springframework.boot.docker.compose.core.RunningService;
import org.springframework.boot.docker.compose.service.connection.DockerComposeConnectionDetailsFactory;
import org.springframework.boot.docker.compose.service.connection.DockerComposeConnectionSource;

import java.net.InetSocketAddress;
import java.util.Optional;

public class MeilisearchDockerComposeConnectionDetailsFactory
        extends DockerComposeConnectionDetailsFactory<MeilisearchConnectionDetails> {

    public MeilisearchDockerComposeConnectionDetailsFactory() {
        super("getmeili/meilisearch", "com.meilisearch.sdk.Client");
    }

    @Override
    public MeilisearchConnectionDetails getDockerComposeConnectionDetails(DockerComposeConnectionSource source) {
        return new MeilisearchDockerComposeConnectionDetails(source.getRunningService());
    }

    public static class MeilisearchDockerComposeConnectionDetails
            extends DockerComposeConnectionDetails
            implements MeilisearchConnectionDetails {

        private final RunningService runningService;

        MeilisearchDockerComposeConnectionDetails(RunningService runningService) {
            super(runningService);
            this.runningService = runningService;
        }

        @Override
        public String address() {
            return Optional.ofNullable(runningService.host()).orElse("localhost") + ":" + runningService.ports().get(7700);
        }

        @Override
        public String key() {
            return runningService.env().get("MEILI_MASTER_KEY");
        }
    }
}

File: spring-meilisearch/src/main/kotlin/com/github/asm0dey/opdsko_spring/meilisearch/MeilisearchDockerComposeConnectionDetailsFactory.kt
package com.github.asm0dey.opdsko_spring.meilisearch

import org.springframework.boot.docker.compose.core.RunningService
import org.springframework.boot.docker.compose.service.connection.DockerComposeConnectionDetailsFactory
import org.springframework.boot.docker.compose.service.connection.DockerComposeConnectionSource
import java.net.InetSocketAddress
import com.meilisearch.sdk.Client as MeilisearchClient

class MeilisearchDockerComposeConnectionDetailsFactory :
    DockerComposeConnectionDetailsFactory<MeilisearchConnectionDetails>(
        "getmeili/meilisearch", MeilisearchClient::class.qualifiedName!!
    ) {

    override fun getDockerComposeConnectionDetails(source: DockerComposeConnectionSource): MeilisearchConnectionDetails {
        return MeilisearchDockerComposeConnectionDetails(source.runningService)
    }

    private class MeilisearchDockerComposeConnectionDetails(val runningService: RunningService) :
        DockerComposeConnectionDetails(runningService),
        MeilisearchConnectionDetails {
        override fun address() = InetSocketAddress(
            runningService.host() ?: "localhost",
            runningService.ports()[7700]
        )

        override fun key() = runningService.env()["MEILI_MASTER_KEY"]!!
    }
}
